File: subsets.py
from typing import List

def subsets(nums: List[int]) -> List[List[int]]:
    # WRITE YOUR BRILLIANT CODE HERE
    def backtrack(nums, curr, result, start):
        if start == len(nums):
            result.append(curr[:])
            return
        
        result.append(curr[:])
        for i in range(start, len(nums)):
            curr.append(nums[i])
            backtrack(nums, curr, result, i + 1)
            curr.pop()
        
        
    result = []
    backtrack(nums, [], result, 0)
    return result


# Solution
# make the binary choice: include or don't include the current number
def subsetsSolution(nums: List[int]) -> List[List[int]]:
    n = len(nums)

    res = []
    def dfs(i, cur):
        if i == n:
            res.append(cur)
            return

        dfs(i + 1, cur + [nums[i]])
        dfs(i + 1, cur)

    dfs(0, [])

    return res


# print(subsets([1,2,3]))
print(subsetsSolution([1,2,3]))

File: subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
// trick is to use i + 1 to pass as the argument instead of start or start + 1
// review the permutation and combination sum to compare the ways how to traverse the tree path while backtracking
 var subsets = function(nums) {
    function backtrack(start, curr, res, nums) {
        if (start > nums.length) return;
        res.push([...curr]);
        for (let i = start; i < nums.length; ++i) {
            curr.push(nums[i]);
            backtrack(i + 1, curr, res, nums);
            curr.pop();
        }
    }
    const res = [];
    backtrack(0, [], res, nums);
    return res;
};

// algomonster solution
// idea: include and don't include
function subsets(nums) {
    const res = [];
    function dfs(i, cur) {
        if (i == nums.length) {
            res.push([...cur]);
            return;
        }
        cur.push(nums[i]);
        dfs(i + 1, cur);
        cur.pop();
        dfs(i + 1, cur);
    }
    dfs(0, []);
    return res;
}

// algomonster alternative approach
// apply pruning
function subsets(nums) {
    const res = [[]];
    function dfs(i, cur) {
        if (i == nums.length) {
            return;
        }
        cur.push(nums[i]);
        res.push([...cur]);
        dfs(i + 1, cur);
        cur.pop();
        dfs(i + 1, cur);
    }
    dfs(0, []);
    return res;
}